audio/out/ao_coreaudio_exclusive.c

```diff
@@ -45,13 +45,18 @@
 #include "internal.h"
 #include "audio/format.h"
 #include "osdep/timer.h"
+#include "osdep/threads.h"
 #include "options/m_option.h"
 #include "common/msg.h"
 #include "audio/out/ao_coreaudio_chmap.h"
 #include "audio/out/ao_coreaudio_properties.h"
 #include "audio/out/ao_coreaudio_utils.h"
 
 struct priv {
+    // This must be put in the front
```

Comment:
why?

Reply:
https://github.com/mpv-player/mpv/blob/c1b8ea82af765dfcbd17a1dcec588d5b02132419/audio/out/ao_coreaudio_utils.c#L478

`ca_change_physical_format_sync` is used by both `ao_coreaudio` and `ao_coreaudio_exclusive`. Their `priv` differ. If we don't put these condvar and mutex in the front, we have to deal with two different struct layout.

Reply:
I see. How about we at least use the struct for that?

``` diff
diff --git a/audio/out/ao_coreaudio.c b/audio/out/ao_coreaudio.c
index 76df770649..99939cbfc6 100644
--- a/audio/out/ao_coreaudio.c
+++ b/audio/out/ao_coreaudio.c
@@ -36,8 +36,7 @@
 
 struct priv {
     // This must be put in the front
-    mp_mutex mutex;
-    mp_cond cond;
+    struct coreaudio_cb_sem sem;
 
     AudioDeviceID device;
     AudioUnit audio_unit;
diff --git a/audio/out/ao_coreaudio_exclusive.c b/audio/out/ao_coreaudio_exclusive.c
index 98edb0b248..47b23781fd 100644
--- a/audio/out/ao_coreaudio_exclusive.c
+++ b/audio/out/ao_coreaudio_exclusive.c
@@ -54,8 +54,7 @@
 
 struct priv {
     // This must be put in the front
-    mp_mutex mutex;
-    mp_cond cond;
+    struct coreaudio_cb_sem sem;
 
     AudioDeviceID device;   // selected device
 
diff --git a/audio/out/ao_coreaudio_utils.c b/audio/out/ao_coreaudio_utils.c
index d8825f3e2a..46db9cac6a 100644
--- a/audio/out/ao_coreaudio_utils.c
+++ b/audio/out/ao_coreaudio_utils.c
@@ -455,11 +455,6 @@ int64_t ca_get_device_latency_ns(struct ao *ao, AudioDeviceID device)
     return MP_TIME_S_TO_NS(latency_frames / sample_rate);
 }
 
-struct cb_sem {
-    mp_mutex mutex;
-    mp_cond cond;
-};
-
 static OSStatus ca_change_format_listener(
     AudioObjectID object, uint32_t n_addresses,
     const AudioObjectPropertyAddress addresses[],
diff --git a/audio/out/ao_coreaudio_utils.h b/audio/out/ao_coreaudio_utils.h
index d11506c059..bd69b5a667 100644
--- a/audio/out/ao_coreaudio_utils.h
+++ b/audio/out/ao_coreaudio_utils.h
@@ -29,6 +29,11 @@
 #include "internal.h"
 #include "osdep/utils-mac.h"
 
+struct coreaudio_cb_sem {
+    mp_mutex mutex;
+    mp_cond cond;
+};
+
 bool check_ca_st(struct ao *ao, int level, OSStatus code, const char *message);
 
 #define CHECK_CA_ERROR_L(label, message) \

```